PayloadPkg/OsLoader/LoadImage.c

```diff
@@ -804,7 +804,32 @@ LoadBootImages (
       Status = GetBootImageFromRawPartition (OsBootOption, LoadedImage);
     }
 
-    DEBUG ((DEBUG_INFO, "LoadBootImage ImageType-%d Image\n", Index));
+    if (EFI_ERROR (Status) && (Index == LoadImageTypeNormal)) {
+      //
+      // Launch Traditional Linux for debugging purpose only
+      //
+      if ((DebugCodeEnabled () || !FeaturePcdGet (PcdVerifiedBootEnabled)) && (FsHandle != NULL)) {
+        // Free loaded images previously, but keep LoadedImagesInfo structure
+        UnloadBootImages ((EFI_HANDLE)(UINTN)LoadedImagesInfo, TRUE);

```

Comment:
This will clear all pre-loaded images. Is this still valid?

Reply:
I just move the code, didn't change the logic. not sure why it is added before.
when loaded IAS image failure before, do we need free anything left?

Reply:
This whole code is for debugging purpose only. If there is no valid boot image found in for loop, this code will clean-up all allocated memory and load linux files for debugging purpose. Do we need to move the code from the original location to inside for loop?

Reply:
For each boot option, it could load preOS, normal OS, A/B support OS, and extra image.
The traditional Linux should be part of the boot option loading when loading normal OS failure to make the loading flow consistent. And we could also test PreOS + traditional linux case with this change.
If we need UnloadBootImages after each loop, we could separate it from traditional linux loading. 
BTW, if UnloadBootImages is needed, should we also check it regardless debug/release build?

Reply:
The new code logic has altered the current code flow now.  We need to think it carefully.

For example:  the old code will fallback to traditional Linux boot no matter which image loading fails.  The new logic will only do that for the 1st image loading failure.   

I agree with Aiden on this.  It might not be necessary to make this change since Traditional Linux is not considered as a normal boot option, instead, it is a fallback option when the official boot option fails.  So it makes sense to me if we handle it outside of the for loop as an error handler.

Reply:
The new logic is expected for the enhancement.
For a simple boot option (only load normal OS), the new logic is same with old logic.
For other boot options (e.g. PreOS + traditional Linux), old logic will change the boot option to only load traditional Linux since it will free PreOS. The new logic fixed it.
I checked UnloadBootImages (), it is ok to keep it there in case there is some memory allocated during IAS image loading.

Reply:
The new logic is not the same as old one because you added condition "Index == LoadImageTypeNormal".

Reply:
BTW,  as discussed, traditional Linux is not considered as a normal boot option. So PreOS+Traditional Linux doesn't seem to be a valid combination in production.   In production, it should be PreOS + Linux Container.

Reply:
That's because traditional Linux will be loaded into Image[LoadImageTypeNormal] below.
For the simple boot option (only load normal OS), other image load will be skipped. so the new logic is same with old logic ( that is, when normal OS loaded failure, it will load traditional Linux).

Reply:
I agree Traditional Linux is not for production, this change doesn't change that.
It would help developer/user to test PreOS with traditional Linux for debug purpose.

Reply:
Give an example, if the old boot option needs to load two images.  The first image loading is successful, but the 2nd image loading fails. With the old code logic, it will try to run Traditional Linux at the end.  But with the new code it will not.

Reply:
I'd prefer to keep original code. It would be risky to add debug code inside main loop. We may need some unit test result with different boot options and boot image combinations. If this is useful to debug a specific case, "Testing" branch would be good place.